lib/modules/embeddings/topic.rb

```diff
@@ -3,54 +3,80 @@
 module DiscourseAi
   module Embeddings
     class Topic
-      def initialize(topic)
-        @topic = topic
-        @embeddings = {}
-      end
-
-      def perform!
+      def generate_and_store_embeddings_for(topic)
         return unless SiteSetting.ai_embeddings_enabled
-        return if DiscourseAi::Embeddings::Models.enabled_models.empty?
+        return if topic.blank? || topic.first_post.blank?
 
-        calculate_embeddings!
-        persist_embeddings! unless @embeddings.empty?
-      end
+        enabled_models = DiscourseAi::Embeddings::Model.enabled_models
+        return if enabled_models.empty?
 
-      def calculate_embeddings!
-        return if @topic.blank? || @topic.first_post.blank?
-
-        DiscourseAi::Embeddings::Models.enabled_models.each do |model|
-          @embeddings[model.name] = send("#{model.provider}_embeddings", model.name)
+        enabled_models.each do |model|
+          embedding = model.generate_embedding(topic.first_post.raw)
+          persist_embedding(topic, model, embedding) if embedding
         end
       end
 
-      def persist_embeddings!
-        @embeddings.each do |model, model_embedding|
-          DiscourseAi::Database::Connection.db.exec(
-            <<~SQL,
-              INSERT INTO topic_embeddings_#{model.underscore} (topic_id, embedding)
-              VALUES (:topic_id, '[:embedding]')
-              ON CONFLICT (topic_id)
-              DO UPDATE SET embedding = '[:embedding]'
-            SQL
-            topic_id: @topic.id,
-            embedding: model_embedding,
-          )
+      def symmetric_semantic_search(model, topic)
```

Comment:
Where is this used?

Reply:
It's used by the `SemanticSuggested` service. I moved the queries there since they only work for topic embeddings.